src/globus_sdk/experimental/session_error/session_error.py

```diff
@@ -0,0 +1,195 @@
+                ", ".join(cls.SUPPORTED_FIELDS.keys())
+            )
+
+        # Enforce the field types
+        for field_name, field_type in cls.SUPPORTED_FIELDS.items():
+            if field_name in param_dict and not isinstance(
+                param_dict[field_name], field_type
+            ):
+                raise ValueError(f"{field_name} must be of type {field_type.__name__}")
+
+        return cls(**param_dict)
+
+    def to_dict(self, include_extra: bool = False) -> t.Dict[str, t.Any]:
+        """
+        Return a session error authorization parameters dictionary.
+
+        :param include_extra: Whether to include stored extra (non-standard) fields in
+            the returned dictionary.
+        :type include_extra: bool
+        """
+        session_error_dict = {}
+
+        # Set any authorization parameters
+        for field in self.SUPPORTED_FIELDS.keys():
+            if getattr(self, field) is not None:
+                session_error_dict[field] = getattr(self, field)
+
+        # Set any extra fields
+        if include_extra:
+            session_error_dict.update(self.extra_fields)
+
+        return session_error_dict
+
+
+class GlobusSessionError(GlobusError):
+    """
+    Represents a Globus Session Error.
+
+    A Session Error is a class of error that is returned by Globus services to
+    indicate that additional authorization is required in order to complete a request
+    and contains information that can be used to request the appropriate authorization.
+
+    :ivar code: The error code for this error.
+    :vartype code: str
+
+    :ivar authorization_parameters: The authorization parameters for this error.
+    :vartype authorization_parameters: GlobusAuthorizationParameters
+
+    :ivar extra_fields: A dictionary of additional fields that were provided. May
+        be used for forward/backward compatibility.
+    :vartype extra_fields: dict
+    """
+
+    def __init__(
+        self,
+        code: str,
+        authorization_parameters: GlobusSessionErrorAuthorizationParameters,
+        **kwargs: t.Any,
+    ):
+        self.code: str = code
+        self.authorization_parameters: GlobusSessionErrorAuthorizationParameters = (
+            authorization_parameters
+        )
+        self.extra_fields = kwargs
+
+    @classmethod
+    def from_dict(cls, error_dict: t.Dict[str, t.Any]) -> "GlobusSessionError":
+        """
+        Instantiate a GlobusSessionError from a dictionary.
+
+        :param error_dict: The dictionary to create the error from.
+        :type error_dict: dict
+        """
+
+        if "code" not in error_dict:
+            raise ValueError("GlobusSessionError must have a code")
+
+        # Enforce that authorization_parameters is in the error_dict and
+        # contains at least one of the fields we expect
+        if "authorization_parameters" not in error_dict:
+            raise ValueError("GlobusSessionError must have authorization_parameters")
```

Comment:
I don't think these are doing quite the same work. A bit contrived but, for instance, if we remove these checks and you called `.from_dict({"code": None, "authorization_parameters": None})` in the context of parsing an error response received from a service that would not raise an error during runtime, so you'd potentially instantiate an invalid session error (and should probably attempt the legacy variants). This is stricter than the constructor.

The other concern is if these fields are absent entirely, the error that is raised would be a TypeError. While we could ignore ValueError and TypeError while trying to parse, I'm a little concerned that bulk catching TypeErrors as well as ValueErrors downstream could cause us to possibly cast too wide of a net. But willing to be convinced otherwise!

Reply:
I was on the fence about the concept of `__init__` doing no validation while `from_dict` does validation until reading these comments.

This is pushing me towards the opinion that allowing the unvalidated path is going to cause confusion -- as it already has.

I vote that we put all the validation into the initializers and that `from_dict` therefore inherits that validation logic. I'll return to the other thread above to offer a thought on how to handle things.